apps/api/src/app/rate-limiting/guards/throttler.guard.ts

```diff
@@ -0,0 +1,182 @@
+import {
+  InjectThrottlerOptions,
+  InjectThrottlerStorage,
+  ThrottlerException,
+  ThrottlerGuard,
+  ThrottlerModuleOptions,
+  ThrottlerOptions,
+  ThrottlerStorage,
+} from '@nestjs/throttler';
+import { CallHandler, ExecutionContext, Injectable, NestInterceptor } from '@nestjs/common';
+import { EvaluateApiRateLimit, EvaluateApiRateLimitCommand } from '../usecases/evaluate-api-rate-limit';
+import { Reflector } from '@nestjs/core';
+import { FeatureFlagCommand, GetIsApiRateLimitingEnabled } from '@novu/application-generic';
+import { ApiRateLimitCategoryEnum, ApiRateLimitCostEnum, IJwtPayload } from '@novu/shared';
+import * as jwt from 'jsonwebtoken';
+import { ThrottlerCost, ThrottlerCategory } from './throttler.decorator';
+
+export enum RateLimitHeaderKeysEnum {
+  RATE_LIMIT_REMAINING = 'RateLimit-Remaining',
+  RATE_LIMIT_LIMIT = 'RateLimit-Limit',
+  RATE_LIMIT_RESET = 'RateLimit-Reset',
+  RATE_LIMIT_POLICY = 'RateLimit-Policy',
+  RETRY_AFTER = 'Retry-After',
+  USER_AGENT = 'User-Agent',
+  AUTHORIZATION = 'Authorization',
+}
+
+export const THROTTLED_EXCEPTION_MESSAGE = 'API rate limit exceeded';
+export const ALLOWED_AUTH_SCHEMES = ['ApiKey'];
```

Comment:
Right now we are not planning to rate limit requests with a Bearer token, because those requests will come from the browser (unless someone spoofs an auth flow outside the browser) and browser-based activity is unlikely to cause extreme server load.

Keen to hear your thoughts on whether you think we should indeed be rate limiting Bearer Auth requests.

Reply:
@djabarovgeorge and I discussed this point, we agreed that unless there is a good reason not to limit Bearer token requests, we should limit their consumption.

However, we will need to add support for a custom cache prefix to be passed to the evaluate rate limit use-case to support multiple token buckets to consume from - it's not desirable for both authentication schemes to share the same bucket as this may be confusing to the user.

Given the additional work required and deviation from the initial scope, we should address this in a separate PR if we think we should limit Bearer token requests.

Reply:
I think we should add a client rate limit with a separate bucket as Richard said in order to guard our system from things like:
* malicious actors that could retrieve the jwt token and automate requests.
* misconfigured systems - could be someone that implements the notification center (jwt token user), which could cause high-traffic requests by bug or misconfigured.

there could be potential other cases.

But this should not be part of this PR/Cycle, we could benefit a lot from the first stage of the rate limit and prioritize the next stage accordingly.